File: cpp/trex_selector_methods/trex_gvs/demo_trex_gvs_01_mc_sim_hastie_en_blocks/demo_trex_gvs_01_mc_sim_hastie_en_blocks.cpp
// ==============================================================================
// demo_trex_gvs_01_mc_sim_hastie_en_blocks.cpp
// ==============================================================================
/**
 * @file demo_trex_gvs_01_mc_sim_hastie_en_blocks.cpp
 *
 * @brief MC simulation: T-Rex+GVS on the Hastie DGP.
 *  This scenario evaluates grouped-signal recovery under strong equicorrelation,
 *  not sparse within-group support identification.
 *  Within-group correlation is controlled through sd_x, with rho = 1 / (1 + sd_x^2).
 *
 *
 * @details
 *  Three parts, each in its own function:
 *    sim_hastie_part_1() — 2-D SNR × rho grid
 *    sim_hastie_part_2() — lambda2_method comparison: CV_1SE_SVD / CV_1SE_CCD
 *    sim_hastie_part_3() — hc_linkage comparison: Single / Complete / Average / WPGMA
 *
 * DGP: three equicorrelated groups of 50 variables each, all active (s = 150),
 *      plus 350 null white-noise variables; p = 500 and n = 200.
 *  Selector: K=20, tFDR=0.1, corr_max=0.98, lambda2=CV_1SE_CCD.
 *  Methods compared in Parts 1–3: TENET, TENET_AUG, TIENET_AUG.
 *  MC: 200 trials per grid point.
 */
// ==============================================================================

#include "trex_gvs_dgps.hpp"
#include "trex_gvs_simulation_utils.hpp"


// ==============================================================================

// Namespace usage
using namespace gvs_demo;

// ==============================================================================
//  Part 1 — 2-D SNR × rho sweep
// ==============================================================================

/**
 * @brief Run a 2-D MC sweep over SNR and rho on the Hastie DGP.
 *
 * Evaluates TENET, TENET_AUG, and TIENET_AUG on an SNR × rho grid, then
 * prints and saves matrix summaries for mean TPP and mean FDP.
 */

static void sim_hastie_part_1(const GVSSimConfig& cfg, const std::string& tag)
{
    const std::vector<double> snr_grid_2d = {
        0.2, 0.5, 1.0, 2.0, 5.0
    };
    const std::vector<double> rho_grid_2d = {
        0.30, 0.50, 0.70, 0.90, 0.95, 0.99
    };

    GVS2DDGPFactory dgp_2d = [&cfg](double snr, double rho, unsigned seed) {
        const double sd_x = std::sqrt((1.0 - rho) / rho);
        return make_hastie_dgp(cfg.n, cfg.p, snr, sd_x, seed);
    };

    auto trex_ctrl = make_gvs_trex_control(cfg.K);

    TRexGVSControlParameter gvs_ctrl_en;
    gvs_ctrl_en.gvs_type       = GVSType::EN;
    gvs_ctrl_en.lambda2_method = LambdaSelectionMethod::CV_1SE_CCD;
    gvs_ctrl_en.corr_max       = cfg.corr_max;
    gvs_ctrl_en.hc_linkage     = hac::LinkageMethod::Single;
    gvs_ctrl_en.en_solver      = ENSolverType::TENET;

    TRexGVSControlParameter gvs_ctrl_en_aug;
    gvs_ctrl_en_aug.gvs_type       = GVSType::EN;
    gvs_ctrl_en_aug.lambda2_method = LambdaSelectionMethod::CV_1SE_CCD;
    gvs_ctrl_en_aug.corr_max       = cfg.corr_max;
    gvs_ctrl_en_aug.hc_linkage     = hac::LinkageMethod::Single;
    gvs_ctrl_en_aug.en_solver      = ENSolverType::TENET_AUG;

    TRexGVSControlParameter gvs_ctrl_ien;
    gvs_ctrl_ien.gvs_type       = GVSType::IEN;
    gvs_ctrl_ien.lambda2_method = LambdaSelectionMethod::CV_1SE_CCD;
    gvs_ctrl_ien.corr_max       = cfg.corr_max;
    gvs_ctrl_ien.hc_linkage     = hac::LinkageMethod::Single;

    cdiag::print_section_header("Part 1/3: 2-D SNR x rho Sweep  |  " + tag);
    std::cout << "  n=" << cfg.n << "  p=" << cfg.p << "  s=150"
              << "  MC=" << cfg.num_MC << "  tFDR=" << fmt_num(cfg.tFDR)
              << "  K=" << cfg.K << "  corr_max=" << fmt_num(cfg.corr_max) << "\n"
              << "  methods: TENET / TENET_AUG / TIENET_AUG  |  lambda2=CV_1SE_CCD"
              << "  linkage=Single\n"
              << "  snr_grid: {";
    for (std::size_t i = 0; i < snr_grid_2d.size(); ++i)
        std::cout << (i ? ", " : "") << fmt_num(snr_grid_2d[i]);
    std::cout << "}\n  rho_grid: {";
    for (std::size_t i = 0; i < rho_grid_2d.size(); ++i)
        std::cout << (i ? ", " : "") << fmt_num(rho_grid_2d[i]);
    std::cout << "}\n\n";

    auto en_2d     = run_gvs_2d_sweep(
        dgp_2d, snr_grid_2d, rho_grid_2d, cfg,
        gvs_ctrl_en,     trex_ctrl, "TENET");

    auto en_aug_2d = run_gvs_2d_sweep(
        dgp_2d, snr_grid_2d, rho_grid_2d, cfg,
        gvs_ctrl_en_aug, trex_ctrl, "TENET_AUG");

    auto ien_2d    = run_gvs_2d_sweep(
        dgp_2d, snr_grid_2d, rho_grid_2d, cfg,
        gvs_ctrl_ien,    trex_ctrl, "TIENET_AUG");

    std::vector<std::string> snr_labels, rho_labels;
    for (double s : snr_grid_2d)
        snr_labels.push_back("snr=" + fmt_num(s));
    for (double r : rho_grid_2d)
        rho_labels.push_back("rho=" + fmt_num(r));

    print_mc_matrix("mean_TPP [TENET]",     snr_labels,
                    rho_labels, en_2d,     true);
    print_mc_matrix("mean_FDP [TENET]",     snr_labels,
                    rho_labels, en_2d,     false);
    print_mc_matrix("mean_TPP [TENET_AUG]", snr_labels,
                    rho_labels, en_aug_2d, true);
    print_mc_matrix("mean_FDP [TENET_AUG]", snr_labels,
                    rho_labels, en_aug_2d, false);
    print_mc_matrix("mean_TPP [TIENET_AUG]",    snr_labels,
                    rho_labels, ien_2d,    true);
    print_mc_matrix("mean_FDP [TIENET_AUG]",    snr_labels,
                    rho_labels, ien_2d,    false);
    save_mc_2d_tables(tag, snr_labels, rho_labels, en_2d,
                      en_aug_2d, ien_2d, cfg, "gvs_" + tag + "_2d");
}



// ==============================================================================
//  Part 2 — lambda2_method comparison
// ==============================================================================

/**
 * @brief Compare lambda2 selection methods at a fixed Hastie operating point.
 *
 * Runs TENET and TIENET_AUG at SNR = 2.0 and rho = 0.7, comparing CV_1SE_SVD against
 * CV_1SE_CCD while holding the rest of the selector settings fixed.
 */

static void sim_hastie_part_2(const GVSSimConfig& cfg, const std::string& tag)
{
    // Fixed operating point: SNR=2.0, rho=0.7.
    // sd_x = sqrt((1 - rho) / rho) = sqrt(0.3 / 0.7)
    constexpr double fixed_snr = 2.0;
    constexpr double fixed_rho = 0.7;
    const double     fixed_sdx = std::sqrt((1.0 - fixed_rho) / fixed_rho);

    GVSDGPFactory dgp_fn = [&cfg, fixed_snr, fixed_sdx](double /*unused*/, unsigned seed) {
        return make_hastie_dgp(cfg.n, cfg.p, fixed_snr, fixed_sdx, seed);
    };

    auto trex_ctrl = make_gvs_trex_control(cfg.K);

    const std::vector<LambdaSelectionMethod> methods = {
        LambdaSelectionMethod::CV_1SE_SVD,
        LambdaSelectionMethod::CV_1SE_CCD
    };
    const std::vector<std::string> method_labels = {"CV_1SE_SVD", "CV_1SE_CCD"};

    cdiag::print_section_header("Part 2/3: lambda2_method Comparison  |  " + tag);
    std::cout << "  n=" << cfg.n << "  p=" << cfg.p << "  s=150"
              << "  MC=" << cfg.num_MC << "  tFDR=" << fmt_num(cfg.tFDR)
              << "  K=" << cfg.K << "\n"
              << "  fixed operating point: SNR=" << fmt_num(fixed_snr)
              << "  rho=" << fmt_num(fixed_rho) << "\n"
              << "  lambda2 methods: CV_1SE_SVD / CV_1SE_CCD"
              << "  |  selectors: TENET, TIENET_AUG  (4 runs)\n\n";

    std::vector<GVSGridPointResult> en_results, ien_results;
    en_results.reserve(methods.size());
    ien_results.reserve(methods.size());

    for (std::size_t i = 0; i < methods.size(); ++i) {
        TRexGVSControlParameter gvs_ctrl_en;
        gvs_ctrl_en.gvs_type       = GVSType::EN;
        gvs_ctrl_en.lambda2_method = methods[i];
        gvs_ctrl_en.corr_max       = cfg.corr_max;
        gvs_ctrl_en.hc_linkage     = hac::LinkageMethod::Single;
        gvs_ctrl_en.en_solver      = ENSolverType::TENET;

        en_results.push_back(
            run_gvs_mc_trials(dgp_fn, 0.0, cfg.num_MC,
                              "TENET+" + method_labels[i],
                              cfg.tFDR, gvs_ctrl_en, trex_ctrl,
                              static_cast<unsigned>(cfg.base_seed)));

        TRexGVSControlParameter gvs_ctrl_ien;
        gvs_ctrl_ien.gvs_type       = GVSType::IEN;
        gvs_ctrl_ien.lambda2_method = methods[i];
        gvs_ctrl_ien.corr_max       = cfg.corr_max;
        gvs_ctrl_ien.hc_linkage     = hac::LinkageMethod::Single;

        ien_results.push_back(
            run_gvs_mc_trials(dgp_fn, 0.0, cfg.num_MC,
                              "TIENET_AUG+" + method_labels[i],
                              cfg.tFDR,
                              gvs_ctrl_ien, trex_ctrl,
                              static_cast<unsigned>(cfg.base_seed)));
    }

    {
        const std::string dir(DEMO_OUTPUT_DIR);
        const std::string stem = "gvs_" + tag + "_lambda2_method";
        std::filesystem::create_directories(dir);
        std::ofstream txt_file(dir + stem + ".txt");
        std::ofstream* fout = txt_file.is_open() ? &txt_file : nullptr;

        auto write = [&](const std::string& s) {
            std::cout << s;
            if (fout) *fout << s;
        };

        {
            std::ostringstream hdr;
            hdr << "\n" << std::string(78, '=') << "\n"
                << "  T-Rex+GVS MC: " << tag << " — lambda2_method comparison"
                << "  (SNR=" << fixed_snr << ", rho=" << fixed_rho
                << ", MC=" << cfg.num_MC << ", tFDR=" << cfg.tFDR << ")\n"
                << std::string(78, '=') << "\n";
            write(hdr.str());
        }

        print_mc_1d_method_table("TENET",  method_labels, en_results,  fout);
        print_mc_1d_method_table("TIENET_AUG", method_labels, ien_results, fout);

        write(std::string(78, '=') + "\n\n");

        if (txt_file.is_open()) {
            txt_file.close();
            std::cout << "[Info] TXT saved to: " << dir + stem + ".txt\n";
        }
        std::ofstream csv(dir + stem + ".csv");
        if (csv.is_open()) {
            csv << "method,metric,lambda2_method,mean,sd\n"
                << std::fixed << std::setprecision(6);
            for (std::size_t i = 0; i < method_labels.size(); ++i) {
                const std::string& lbl = method_labels[i];
                csv << "TENET,mean_FDP,"  << lbl << "," << en_results[i].mean_fdp  << "," << en_results[i].sd_fdp  << "\n";
                csv << "TENET,mean_TPP,"  << lbl << "," << en_results[i].mean_tpp  << "," << en_results[i].sd_tpp  << "\n";
                csv << "TIENET_AUG,mean_FDP," << lbl << "," << ien_results[i].mean_fdp << "," << ien_results[i].sd_fdp << "\n";
                csv << "TIENET_AUG,mean_TPP," << lbl << "," << ien_results[i].mean_tpp << "," << ien_results[i].sd_tpp << "\n";
            }
            std::cout << "[Info] CSV saved to: " << dir + stem + ".csv\n\n";
        }
    }
}


// ==============================================================================
//  Part 3 — hc_linkage comparison
// ==============================================================================

/**
 * @brief Compare HAC linkage rules at a fixed Hastie operating point.
 *
 * Runs TENET and TIENET_AUG at SNR = 2.0 and rho = 0.7, comparing Single, Complete,
 * Average, and WPGMA linkage with lambda2 fixed at CV_1SE_CCD.
 */

static void sim_hastie_part_3(const GVSSimConfig& cfg, const std::string& tag)
{
    // Same fixed operating point as Part 2: SNR=2.0, rho=0.7.
    // CV_1SE_CCD used throughout to isolate the linkage effect from lambda selection.
    constexpr double fixed_snr = 2.0;
    constexpr double fixed_rho = 0.7;
    const double     fixed_sdx = std::sqrt((1.0 - fixed_rho) / fixed_rho);

    GVSDGPFactory dgp_fn = [&cfg, fixed_snr, fixed_sdx](double /*unused*/, unsigned seed) {
        return make_hastie_dgp(cfg.n, cfg.p, fixed_snr, fixed_sdx, seed);
    };

    auto trex_ctrl = make_gvs_trex_control(cfg.K);

    const std::vector<hac::LinkageMethod> linkages = {
        hac::LinkageMethod::Single,
        hac::LinkageMethod::Complete,
        hac::LinkageMethod::Average,
        hac::LinkageMethod::WPGMA
    };
    const std::vector<std::string> linkage_labels = {
        "Single", "Complete", "Average", "WPGMA"};

    cdiag::print_section_header("Part 3/3: hc_linkage Comparison  |  " + tag);
    std::cout << "  n=" << cfg.n << "  p=" << cfg.p << "  s=150"
              << "  MC=" << cfg.num_MC << "  tFDR=" << fmt_num(cfg.tFDR)
              << "  K=" << cfg.K << "  lambda2=CV_1SE_CCD\n"
              << "  fixed operating point: SNR=" << fmt_num(fixed_snr)
              << "  rho=" << fmt_num(fixed_rho) << "\n"
              << "  linkages: Single / Complete / Average / WPGMA"
              << "  |  selectors: TENET, TIENET_AUG  (8 runs)\n\n";

    std::vector<GVSGridPointResult> en_results, ien_results;
    en_results.reserve(linkages.size());
    ien_results.reserve(linkages.size());

    for (std::size_t i = 0; i < linkages.size(); ++i) {
        TRexGVSControlParameter gvs_ctrl_en;
        gvs_ctrl_en.gvs_type       = GVSType::EN;
        gvs_ctrl_en.lambda2_method = LambdaSelectionMethod::CV_1SE_CCD;
        gvs_ctrl_en.corr_max       = cfg.corr_max;
        gvs_ctrl_en.hc_linkage     = linkages[i];
        gvs_ctrl_en.en_solver      = ENSolverType::TENET;

        en_results.push_back(
            run_gvs_mc_trials(dgp_fn, 0.0, cfg.num_MC,
                              "TENET+" + linkage_labels[i],
                              cfg.tFDR, gvs_ctrl_en, trex_ctrl,
                              static_cast<unsigned>(cfg.base_seed)));

        TRexGVSControlParameter gvs_ctrl_ien;
        gvs_ctrl_ien.gvs_type       = GVSType::IEN;
        gvs_ctrl_ien.lambda2_method = LambdaSelectionMethod::CV_1SE_CCD;
        gvs_ctrl_ien.corr_max       = cfg.corr_max;
        gvs_ctrl_ien.hc_linkage     = linkages[i];

        ien_results.push_back(
            run_gvs_mc_trials(dgp_fn, 0.0, cfg.num_MC,
                              "TIENET_AUG+" + linkage_labels[i],
                              cfg.tFDR, gvs_ctrl_ien, trex_ctrl,
                              static_cast<unsigned>(cfg.base_seed)));
    }

    {
        const std::string dir(DEMO_OUTPUT_DIR);
        const std::string stem = "gvs_" + tag + "_hc_linkage";
        std::filesystem::create_directories(dir);
        std::ofstream txt_file(dir + stem + ".txt");
        std::ofstream* fout = txt_file.is_open() ? &txt_file : nullptr;

        auto write = [&](const std::string& s) {
            std::cout << s;
            if (fout) *fout << s;
        };

        {
            std::ostringstream hdr;
            hdr << "\n" << std::string(78, '=') << "\n"
                << "  T-Rex+GVS MC: " << tag << " — hc_linkage comparison"
                << "  (SNR=" << fixed_snr << ", rho=" << fixed_rho
                << ", MC=" << cfg.num_MC << ", tFDR=" << cfg.tFDR << ")\n"
                << std::string(78, '=') << "\n";
            write(hdr.str());
        }

        print_mc_1d_method_table("TENET",  linkage_labels, en_results,  fout);
        print_mc_1d_method_table("TIENET_AUG", linkage_labels, ien_results, fout);

        write(std::string(78, '=') + "\n\n");

        if (txt_file.is_open()) {
            txt_file.close();
            std::cout << "[Info] TXT saved to: " << dir + stem + ".txt\n";
        }
        std::ofstream csv(dir + stem + ".csv");
        if (csv.is_open()) {
            csv << "method,metric,hc_linkage,mean,sd\n"
                << std::fixed << std::setprecision(6);
            for (std::size_t i = 0; i < linkage_labels.size(); ++i) {
                const std::string& lbl = linkage_labels[i];
                csv << "TENET,mean_FDP,"  << lbl << "," << en_results[i].mean_fdp  << "," << en_results[i].sd_fdp  << "\n";
                csv << "TENET,mean_TPP,"  << lbl << "," << en_results[i].mean_tpp  << "," << en_results[i].sd_tpp  << "\n";
                csv << "TIENET_AUG,mean_FDP," << lbl << "," << ien_results[i].mean_fdp << "," << ien_results[i].sd_fdp << "\n";
                csv << "TIENET_AUG,mean_TPP," << lbl << "," << ien_results[i].mean_tpp << "," << ien_results[i].sd_tpp << "\n";
            }
            std::cout << "[Info] CSV saved to: " << dir + stem + ".csv\n\n";
        }
    }
}


// ==============================================================================
//  main
// ==============================================================================

int main()
{

    // ==========================================================================
    //  Simulation parameters
    // ==========================================================================

    std::cout.setf(std::ios::unitbuf);
    omp_set_num_threads(6);
    std::cout << "Running with " << omp_get_max_threads() << " threads\n\n";

    GVSSimConfig cfg;
    cfg.n         = 200;
    cfg.p         = 500;
    cfg.sd_x      = 0.1;    // sqrt(0.01) => rho ≈ 0.99
    cfg.tFDR      = 0.1;
    cfg.K         = 20;
    cfg.num_MC    = 200;
    cfg.base_seed = 2026;
    cfg.corr_max  = 0.98;
    cfg.snr       = 2.0;    // unused here; Parts 2-3 fix SNR = 2.0 locally

    const std::string scenario_tag = "Hastie";

    // ==========================================================================
    //  Run simulations
    // ==========================================================================

    if (true) sim_hastie_part_1(cfg, scenario_tag);
    if (true) sim_hastie_part_2(cfg, scenario_tag);
    if (true) sim_hastie_part_3(cfg, scenario_tag);

    std::cout << "\n" << scenario_tag << " GVS MC simulations complete.\n";
    return 0;
}

File: Python/trex_selector_methods/trex_gvs/demo_trex_gvs_01_mc_sim_hastie_en_blocks/demo_trex_gvs_01_mc_sim_hastie_en_blocks.py
# ==============================================================================
# demo_trex_gvs_01_mc_sim_hastie_en_blocks.py
# ==============================================================================
#
# MC simulation: T-Rex+GVS on the Hastie DGP.
# This scenario evaluates grouped-signal recovery under strong equicorrelation,
# not sparse within-group support identification. Within-group correlation is
# controlled through sd_x, with rho = 1 / (1 + sd_x^2).
#
# Mirrors cpp/trex_selector_methods/trex_gvs/demo_trex_gvs_01_mc_sim_hastie_en_blocks.
#
# Three parts, each in its own function:
#   sim_hastie_part_1() — 2-D SNR x rho grid
#   sim_hastie_part_2() — lambda2_method comparison: CV_1SE_SVD / CV_1SE_CCD
#   sim_hastie_part_3() — hc_linkage comparison: Single / Complete / Average / WPGMA
#
# DGP: three equicorrelated groups of 50 variables each, all active (s = 150),
#      plus 350 null white-noise variables; p = 500 and n = 200.
# Selector: K=20, tFDR=0.1, corr_max=0.98, lambda2=CV_1SE_CCD.
# Methods compared: TENET, TENET_AUG, TIENET_AUG.
# MC: 200 trials per grid point (the full run takes hours — reduce num_MC and
#     the grids for a quick smoke test). Optional first CLI argument sets the
#     number of parallel worker processes (default min(6, cpu_count)).
# Indices are 0-based (Python convention); the R counterpart is 1-based.
# ==============================================================================

import math
import os
import sys

_THIS_DIR = os.path.dirname(os.path.abspath(__file__))
_PARENT_DIR = os.path.dirname(_THIS_DIR)
for _p in (_THIS_DIR, _PARENT_DIR):
    if _p not in sys.path:
        sys.path.insert(0, _p)

from gvs_sim_common import (N_CORES, fmt_num, print_mc_matrix,
                            print_section_header, run_gvs_2d_sweep,
                            run_gvs_mc_trials, save_mc_1d_method_tables,
                            save_mc_2d_tables)

N_WORKERS = int(sys.argv[1]) if len(sys.argv) > 1 and sys.argv[1].isdigit() else N_CORES

OUT_DIR = os.path.join(_THIS_DIR, "simulation_results", "data")

# Simulation parameters (mirrors the cpp GVSSimConfig block in main()).
CFG = dict(
    n=200, p=500,
    sd_x=0.1,           # sqrt(0.01) => rho ~ 0.99 (Parts fix rho locally)
    tFDR=0.1, K=20,
    num_MC=200,
    base_seed=2026,
    corr_max=0.98,
    snr=2.0,            # unused here; Parts 2-3 fix SNR = 2.0 locally
)

SCENARIO_TAG = "Hastie"

RUN_PART_1 = True
RUN_PART_2 = True
RUN_PART_3 = True


# ==============================================================================
#  Part 1 — 2-D SNR x rho sweep
# ==============================================================================

def sim_hastie_part_1(cfg, tag):
    """2-D MC sweep over SNR and rho: TENET / TENET_AUG / TIENET_AUG."""
    snr_grid_2d = [0.2, 0.5, 1.0, 2.0, 5.0]
    rho_grid_2d = [0.30, 0.50, 0.70, 0.90, 0.95, 0.99]

    def cell_kwargs(snr, rho):
        sd_x = math.sqrt((1.0 - rho) / rho)
        return dict(n=cfg["n"], p=cfg["p"], snr=snr, sd_x=sd_x)

    print_section_header(f"Part 1/3: 2-D SNR x rho Sweep  |  {tag}")
    print(f"  n={cfg['n']}  p={cfg['p']}  s=150  MC={cfg['num_MC']}"
          f"  tFDR={fmt_num(cfg['tFDR'])}  K={cfg['K']}"
          f"  corr_max={fmt_num(cfg['corr_max'])}\n"
          "  methods: TENET / TENET_AUG / TIENET_AUG  |  lambda2=CV_1SE_CCD"
          "  linkage=Single\n"
          "  snr_grid: {" + ", ".join(fmt_num(s) for s in snr_grid_2d) + "}\n"
          "  rho_grid: {" + ", ".join(fmt_num(r) for r in rho_grid_2d) + "}\n")

    en_2d = run_gvs_2d_sweep("dgp_hastie_snr", cell_kwargs, snr_grid_2d,
                             rho_grid_2d, cfg, "EN", "TENET", "TENET",
                             max_workers=N_WORKERS)
    en_aug_2d = run_gvs_2d_sweep("dgp_hastie_snr", cell_kwargs, snr_grid_2d,
                                 rho_grid_2d, cfg, "EN", "TENET_AUG",
                                 "TENET_AUG", max_workers=N_WORKERS)
    ien_2d = run_gvs_2d_sweep("dgp_hastie_snr", cell_kwargs, snr_grid_2d,
                              rho_grid_2d, cfg, "IEN", "TENET", "TIENET_AUG",
                              max_workers=N_WORKERS)

    snr_labels = [f"snr={fmt_num(s)}" for s in snr_grid_2d]
    rho_labels = [f"rho={fmt_num(r)}" for r in rho_grid_2d]

    print_mc_matrix("mean_TPP [TENET]", snr_labels, rho_labels, en_2d, True)
    print_mc_matrix("mean_FDP [TENET]", snr_labels, rho_labels, en_2d, False)
    print_mc_matrix("mean_TPP [TENET_AUG]", snr_labels, rho_labels, en_aug_2d, True)
    print_mc_matrix("mean_FDP [TENET_AUG]", snr_labels, rho_labels, en_aug_2d, False)
    print_mc_matrix("mean_TPP [TIENET_AUG]", snr_labels, rho_labels, ien_2d, True)
    print_mc_matrix("mean_FDP [TIENET_AUG]", snr_labels, rho_labels, ien_2d, False)
    save_mc_2d_tables(tag, snr_labels, rho_labels, en_2d, en_aug_2d, ien_2d,
                      cfg, OUT_DIR, f"gvs_{tag}_2d")


# ==============================================================================
#  Part 2 — lambda2_method comparison
# ==============================================================================

def sim_hastie_part_2(cfg, tag):
    """Compare lambda2 selection methods (CV_1SE_SVD vs CV_1SE_CCD) for TENET
    and TIENET_AUG at the fixed operating point SNR = 2.0, rho = 0.7."""
    fixed_snr, fixed_rho = 2.0, 0.7
    fixed_sdx = math.sqrt((1.0 - fixed_rho) / fixed_rho)
    dgp_kwargs = dict(n=cfg["n"], p=cfg["p"], snr=fixed_snr, sd_x=fixed_sdx)

    methods = ["CV_1SE_SVD", "CV_1SE_CCD"]

    print_section_header(f"Part 2/3: lambda2_method Comparison  |  {tag}")
    print(f"  n={cfg['n']}  p={cfg['p']}  s=150  MC={cfg['num_MC']}"
          f"  tFDR={fmt_num(cfg['tFDR'])}  K={cfg['K']}\n"
          f"  fixed operating point: SNR={fmt_num(fixed_snr)}"
          f"  rho={fmt_num(fixed_rho)}\n"
          "  lambda2 methods: CV_1SE_SVD / CV_1SE_CCD"
          "  |  selectors: TENET, TIENET_AUG  (4 runs)\n")

    en_results, ien_results = [], []
    for lm in methods:
        en_results.append(run_gvs_mc_trials(
            "dgp_hastie_snr", dgp_kwargs, cfg["num_MC"], f"TENET+{lm}",
            cfg["tFDR"], cfg["K"], "EN", cfg["corr_max"],
            en_solver="TENET", lambda2_method=lm,
            base_seed=cfg["base_seed"], max_workers=N_WORKERS))
        ien_results.append(run_gvs_mc_trials(
            "dgp_hastie_snr", dgp_kwargs, cfg["num_MC"], f"TIENET_AUG+{lm}",
            cfg["tFDR"], cfg["K"], "IEN", cfg["corr_max"],
            lambda2_method=lm,
            base_seed=cfg["base_seed"], max_workers=N_WORKERS))

    save_mc_1d_method_tables(
        tag, "lambda2_method", "lambda2_method", methods, en_results,
        ien_results, f"SNR={fmt_num(fixed_snr)}, rho={fmt_num(fixed_rho)}",
        cfg, OUT_DIR, f"gvs_{tag}_lambda2_method")


# ==============================================================================
#  Part 3 — hc_linkage comparison
# ==============================================================================

def sim_hastie_part_3(cfg, tag):
    """Compare HAC linkage rules (Single / Complete / Average / WPGMA) for
    TENET and TIENET_AUG at the same fixed operating point as Part 2, with
    lambda2 fixed at CV_1SE_CCD to isolate the linkage effect."""
    fixed_snr, fixed_rho = 2.0, 0.7
    fixed_sdx = math.sqrt((1.0 - fixed_rho) / fixed_rho)
    dgp_kwargs = dict(n=cfg["n"], p=cfg["p"], snr=fixed_snr, sd_x=fixed_sdx)

    linkages = ["Single", "Complete", "Average", "WPGMA"]

    print_section_header(f"Part 3/3: hc_linkage Comparison  |  {tag}")
    print(f"  n={cfg['n']}  p={cfg['p']}  s=150  MC={cfg['num_MC']}"
          f"  tFDR={fmt_num(cfg['tFDR'])}  K={cfg['K']}  lambda2=CV_1SE_CCD\n"
          f"  fixed operating point: SNR={fmt_num(fixed_snr)}"
          f"  rho={fmt_num(fixed_rho)}\n"
          "  linkages: Single / Complete / Average / WPGMA"
          "  |  selectors: TENET, TIENET_AUG  (8 runs)\n")

    en_results, ien_results = [], []
    for lk in linkages:
        en_results.append(run_gvs_mc_trials(
            "dgp_hastie_snr", dgp_kwargs, cfg["num_MC"], f"TENET+{lk}",
            cfg["tFDR"], cfg["K"], "EN", cfg["corr_max"],
            hc_linkage=lk, en_solver="TENET",
            base_seed=cfg["base_seed"], max_workers=N_WORKERS))
        ien_results.append(run_gvs_mc_trials(
            "dgp_hastie_snr", dgp_kwargs, cfg["num_MC"], f"TIENET_AUG+{lk}",
            cfg["tFDR"], cfg["K"], "IEN", cfg["corr_max"],
            hc_linkage=lk,
            base_seed=cfg["base_seed"], max_workers=N_WORKERS))

    save_mc_1d_method_tables(
        tag, "hc_linkage", "hc_linkage", linkages, en_results, ien_results,
        f"SNR={fmt_num(fixed_snr)}, rho={fmt_num(fixed_rho)}",
        cfg, OUT_DIR, f"gvs_{tag}_hc_linkage")


# ==============================================================================
#  main
# ==============================================================================

if __name__ == "__main__":
    print(f"Running with {N_WORKERS} worker processes\n")

    if RUN_PART_1:
        sim_hastie_part_1(CFG, SCENARIO_TAG)
    if RUN_PART_2:
        sim_hastie_part_2(CFG, SCENARIO_TAG)
    if RUN_PART_3:
        sim_hastie_part_3(CFG, SCENARIO_TAG)

    print(f"\n{SCENARIO_TAG} GVS MC simulations complete.")
